tests/cql_query_test.cc

```diff
@@ -365,6 +365,34 @@ SEASTAR_TEST_CASE(test_in_clause_validation) {
     });
 }
 
+SEASTAR_THREAD_TEST_CASE(test_in_clause_cartesian_product_limits) {
+    do_with_cql_env_thread([] (cql_test_env& e) {
+        e.execute_cql("CREATE TABLE tab1 (pk1 int, pk2 int, PRIMARY KEY ((pk1, pk2)))").get();
+
+        // 100 partitions, should pass
+        e.execute_cql("SELECT * FROM tab1 WHERE pk1 IN (0, 1, 2, 3, 4, 5, 6, 7, 8, 9)"
+                "                           AND pk2 IN (0, 1, 2, 3, 4, 5, 6, 7, 8, 9)").get();
+        // 110 partitions, should fail
+        BOOST_REQUIRE_THROW(
+                e.execute_cql("SELECT * FROM tab1 WHERE pk1 IN (0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10)"
+                        "                          AND pk2 IN (0, 1, 2, 3, 4, 5, 6, 7, 8, 9)").get(),
+                        std::runtime_error);
+
+        e.execute_cql("CREATE TABLE tab2 (pk1 int, ck1 int, ck2 int, PRIMARY KEY (pk1, ck1, ck2))").get();
+
+        // 100 clustering rows, should pass
+        e.execute_cql("SELECT * FROM tab2 WHERE pk1 = 1"
+                "                          AND ck1 IN (0, 1, 2, 3, 4, 5, 6, 7, 8, 9)"
+                "                          AND ck2 IN (0, 1, 2, 3, 4, 5, 6, 7, 8, 9)").get();
+        // 110 clustering rows, should fail
+        BOOST_REQUIRE_THROW(
+                e.execute_cql("SELECT * FROM tab2 WHERE pk1 = 1"
+                        "                           AND ck1 IN (0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10)"
+                        "                           AND ck2 IN (0, 1, 2, 3, 4, 5, 6, 7, 8, 9)").get(),
+                        std::runtime_error);
+    }).get();
+}
+
```

Comment:
One of the issues we had that led to the development of this patch was a single partition key being restricted.
Last we spoke I remember you told me that this would generate the same code paths in the end, despite not being a cartesian product.

In that case shouldn't we add a test case?

My fear is that someone may in the future change that and dissociate the code paths in which case queries that were invalid will magically become valid, despite having the same number of restrictions

Reply:
Sure.

Reply:
Added 4 cases: {pass, fail} x {one partition key, one clustering key}